React/Testing best practices.md

```diff
@@ -237,78 +210,215 @@ Query priorities ([more info](https://testing-library.com/docs/guide-which-query
+
+1. Root `describe` must be named after component/page/hook/util we're testing
+2. Nested `describe`s must have `when` prefix (to indicate specific scenarios)
+3. Description of `it` must be a use case sentence
+
+An example:
+
+```tsx
+  describe('useAuth', () => {
+    it('should throw context error', () => {
+      // ...
+    });
+    it('should toggle loading state', () => {
+      // ..
+    });
+
+    describe('when user exists', () => {
+      it('should user object', () => {
+        // ...
+      });
+      it('should log out user', () => {
+        // ...
+      });
+    });
+
+    describe('when user does not exist', () => {
+      it('should return guest user', () => {
+        // ...
+      });
+      it('should destroy session on window close', () => {
+        // ...
+      });
+    });
+  });
+```
+
+### Basic test example
+
+Base Component:
 
 ```tsx
-const Button: FC<ButtonProps> = ({ children, ...rest }) => (
-  <button {...rest}>{children}</button>
+const Button: FC<ButtonProps> = (props) => (
+  <button {...props} />
 );
 ```
 
 Test:
 
 ```tsx
 describe("Button", () => {
-  it("Is rendering", () => {
-    const buttonText = "click here";
-
-    const { getByText } = render(<Button>{buttonText}</Button>);
-
-    expect(getByText(buttonText)).toBeDefined();
-  });
-
-  it("Is onClick called", () => {
+  // or something more meaningful
+  it("should call onClick handler", () => {
     const buttonText = "click here";
     const testOnClick = jest.fn();
 
-    const { getByText } = render(
+    render(
       <Button onClick={testOnClick}>{buttonText}</Button>
     );
 
-    userEvent.click(getByText(buttonText));
+    user.click(screen.getByText(buttonText));
 
     expect(testOnClick).toBeCalledTimes(1);
   });
 });
 ```
 
-### Components that uses a base component test example
+**Components that uses a base component test example**
 
 Component:
 
 ```tsx
 import { Button } from "components/Button";
 
-const UserCard: FC<UserCardProps> = ({ label }) => (
+const UserCard: FC<UserCardProps> = ({ title }) => (
   <Card>
-    <h3>{label}</h3>
+    <h3>{title}</h3>
     <Button>click</Button>
   </Card>
 );
 ```
 
 Test:
 
+Here, we mock our `Button` component implementation because we don't care about it here, we only care about the specifics of the `UserCard` component. This is especially useful when you don't want to generate a large tree inside your test for components that don't have any impact on the actual test.
+
 ```jsx
+import { screen } from '@testing-library/react';
+
 jest.mock("components/button");
 (Button as jest.Mock).mockReturnValue(<button />);
 
 describe("UserCard", () => {
-  it("Is rendering", () => {
+  it("should display the correct title", () => {
     const username = "Test User";
 
-    const { getByText } = render(<UserCard label={username} />);
+    render(<UserCard title={username} />);
 
-    expect(getByText(username)).toBeDefined();
+    expect(screen.(username)).toBeDefined();
```

Comment:
```suggestion
    expect(screen.getByText(username)).toBeDefined();
```

As we talked in person, please add a paragraph on using screen over destructuring the render method return object.